src/oss-find-squats/FindSquatsTool.cs

```diff
@@ -0,0 +1,145 @@
+                                                     .Product;
+
+                runItem.Tool.Driver.Version = entryAssembly.GetCustomAttribute<AssemblyInformationalVersionAttribute>()?
+                                                    .InformationalVersion;
+            }
+
+            var descriptors = new List<ReportingDescriptor>();
+
+            foreach (var mutation in gen.Mutations)
+            {
+                ReportingDescriptor sarifRule = new ReportingDescriptor();
+                sarifRule.Name = mutation.Method.Name;
+                sarifRule.DefaultConfiguration = new ReportingConfiguration() { Enabled = true };
+            }
+
+            runItem.Tool.Driver.Rules = descriptors;
+
+            var sarifResults = new List<Result>();
+
+            foreach (var target in opts.Targets ?? Array.Empty<string>())
+            {
+                var purl = new PackageURL(target);
+                if (purl.Name is not null && purl.Type is not null)
+                {
+                    var manager = ProjectManagerFactory.CreateProjectManager(purl, null);
+                    if (manager is not null)
+                    {
+                        var mutationsDict = gen.Mutate(purl.Name);
+
+                        foreach ((var candidate, var rules) in mutationsDict)
+                        {
+                            Logger.Info($"Checking {candidate}.");
+                            var candidatePurl = new PackageURL(purl.Type, candidate);
+                            var versions = await manager.EnumerateVersions(candidatePurl);
+
+                            if (versions.Any())
+                            {
+                                Logger.Warn($"{candidate} package exists. Potential squat.");
+                                Result res = new Result();
+                                res.Locations = new List<Location>();
+                                Location loc = new Location();
+
+                                loc.LogicalLocation = new LogicalLocation()
+                                {
+                                    FullyQualifiedName = purl.ToString()
+                                };
+                                res.Locations.Add(loc);
+                                var rulesString = string.Join(',', rules);
+                                var message = $"Potential Squat candidate {candidate} detected. Generated by {rulesString}";
+                                res.Message = new Message(message, null, rulesString, null, null);
+                                res.Kind = ResultKind.Review;
+                                sarifResults.Add(res);
+                            }
+                        }
+                    }
+                }
+            }
+
+            runItem.Results = sarifResults;
+
+            sarifLog.Runs = new List<Run>();
+            sarifLog.Runs.Add(runItem);
```

Comment:
Checkout Shared\Helpers\OutputBuilder\SarifOutputBuilder.cs.
There are several helper functions there to do many of the sarif level things done here, and also to multiplex between text and sarif. There are examples of using the OutputBuilder in most of the tools.